crates/matrix-sdk-ui/src/timeline/read_receipts.rs

```diff
@@ -279,11 +396,53 @@ impl TimelineInnerStateTransaction<'_> {
 
         self.meta.read_receipts.maybe_update_read_receipt(
             full_receipt,
-            None,
             is_own_event,
+            &self.meta.all_events,
             &mut self.items,
         );
     }
+
+    /// Update the read receipts on the event with the given event ID and the
+    /// previous visible event because of a visibility change.
+    pub(super) fn maybe_update_read_receipts_of_prev_event(&mut self, event_id: &EventId) {
+        // Find the previous visible event, if there is one.
+        let Some((prev_item_pos, prev_event_item)) = self
+            .all_events
+            .iter()
+            .rev()
+            // Find the event item.
+            .skip_while(|meta| meta.event_id != event_id)
+            // Go past the event item.
+            .skip(1)
+            // Find the first visible item.
+            .find(|meta| meta.visible)
+            .and_then(|meta| rfind_event_by_id(&self.items, &meta.event_id))
+        else {
+            return;
+        };
+
+        let prev_event_item_id = prev_event_item.internal_id;
+        let mut prev_event_item = prev_event_item.clone();
+
+        let Some(remote_prev_event_item) = prev_event_item.as_remote_mut() else {
+            warn!("loading read receipts for a local item, this should not be possible");
+            return;
+        };
+
+        let read_receipts = self.read_receipts.read_receipts_for_event(
+            &remote_prev_event_item.event_id,
+            &self.all_events,
+            false,
+        );
+
+        // If the count did not change, the receipts did not change either.
+        if read_receipts.len() == remote_prev_event_item.read_receipts.len() {
+            return;
+        }
```

Comment:
Because we always process one receipt at a time?

Reply:
Yes, because every change happens sequentially and this particular change doesn't trigger any change in the read receipts "real" location.

Reply:
I don't fully understand but I trust your judgement 😅